src/tools/kdb/get.c

```diff
@@ -0,0 +1,225 @@
+		CLI_PRINT (CLI_LOG_VERBOSE, "The key is %s.\n", BOLD (binSize == 0 ? "null" : "binary"));
+		const uint8_t * data = (const uint8_t *) keyValue (found);
+		for (int position = 0; position < keyGetValueSize (found); position++)
+		{
+			CLI_PRINT (CLI_LOG_NONE, "\\x%02x", data[position]);
+		}
+	}
+	else
+	{
+		CLI_PRINT (CLI_LOG_NONE, "%s", keyString (found));
+	}
+
+cleanup:
+	if (!noNewLine)
+	{
+		printf ("\n");
+	}
+	keyDel (parentKey);
+	keyDel (toLookUp);
+	ksDel (searchIn);
+
+	RETURN (ret)
+}
+
+void printOptions (elektraLookupFlags options)
+{
+	if (options & KDB_O_SPEC) printf ("KDB_O_SPEC ");
+	if (options & KDB_O_CREATE) printf ("KDB_O_CREATE ");
+	if (options & KDB_O_NOCASCADING) printf ("KDB_O_NOCASCADING ");
+	if (options & KDB_O_NOSPEC) printf ("KDB_O_NOSPEC ");
+	if (options & KDB_O_NODEFAULT) printf ("KDB_O_NODEFAULT ");
+	if (options & KDB_O_CALLBACK) printf ("KDB_O_CALLBACK");
+}
+
+const char * getCascadingName (const char * str)
+{
+	if (str == NULL) return "/";
+	const char * r = strchr (str, '/');
+	return r == NULL ? "/" : r;
+}
+
+Key * warnOnMeta (ELEKTRA_UNUSED KeySet * ks, ELEKTRA_UNUSED Key * key, Key * found, elektraLookupFlags options)
+{
+	if (found != NULL && strncmp (keyName (found), "spec:/", 6) == 0 && options == KDB_O_CALLBACK)
+	{
+		const Key * meta = keyGetMeta (found, "context");
+		if (meta != NULL)
+		{
+			ELEKTRA_ADD_RESOURCE_WARNINGF (
+				key, "%s is context dependent, shown result might be wrong, -v shows you the trace to the key\n",
+				keyName (found));
+		}
+	}
+	return found;
+}
+
+
+Key * printTrace (ELEKTRA_UNUSED KeySet * ks, Key * key, Key * found, elektraLookupFlags options)
+{
+	warnOnMeta (ks, key, found, options);
+	const char * lastKeyName = keyValue (keyGetMeta (key, "callback/print_trace/last_key_name"));
+	const char * name = keyName (key);
+	const char * rawDepth = keyValue (keyGetMeta (key, "callback/print_trace/depth"));
+
+	int depth = rawDepth == NULL ? 0 : atoi (rawDepth);
+	for (int i = 0; i < depth; ++i)
+	{
+		printf (" ");
+	}
+
+	printf ("searching %s%s", (name[0] == '/' ? "default of spec" : ""), name);
+	printf (", found: %s", (found != NULL ? keyName (found) : "<nothing>"));
+	if (options)
+	{
+		printf (", options: ");
+		printOptions (options);
+	}
+	printf ("\n");
+
+	int newDepth = depth;
+	if (elektraStrNCmp (name, "spec:/", 6) == 0 && (options & KDB_O_CALLBACK))
+	{
+		newDepth += 4;
+	}
+	else if (elektraStrCmp (getCascadingName (lastKeyName), getCascadingName (name)) != 0)
+	{
+		newDepth = depth != 0 ? depth - 2 : depth;
+	}
+	if (newDepth != depth)
+	{
+		char buff[11];
+		snprintf (buff, 11, "%d", newDepth);
+		keySetMeta (key, "callback/print_trace/depth", buff);
+	}
+	keySetMeta (key, "callback/print_trace/last_key_name", name);
+	return found;
+}
+
+void setCallback (Key * key, Key * (*f) (KeySet * ks, Key * key, Key * found, elektraLookupFlags flags))
+{
+	union
+	{
+		Key * (*f) (KeySet * ks, Key * key, Key * found, elektraLookupFlags flags);
```

Comment:
```suggestion
		lookup_fn f;
```

Please use a `typedef` for this ungodly type:

```c
typedef Key * (*lookup_fn) (KeySet * ks, Key * key, Key * found, elektraLookupFlags flags);
```

Reply:
done